apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs

```diff
@@ -413,26 +418,81 @@ private LogEntry BuildLogEntry(LogEventInfo loggingEvent)
                 LogName = _logName,
                 Resource = _resource,
             };
 
-            foreach (var combinedProperty in GetAllProperties(loggingEvent))
+            if (SendJsonPayload)
             {
-                if (string.IsNullOrEmpty(combinedProperty.Key))
-                    continue;
+                var jsonStruct = new Struct();
+                jsonStruct.Fields.Add("message", Value.ForString(RenderLogEvent(Layout, loggingEvent)));
 
-                try
+                var propertiesStruct = new Struct();
+                jsonStruct.Fields.Add("properties", Value.ForStruct(propertiesStruct));
+
+                foreach (var combinedProperty in GetAllProperties(loggingEvent))
                 {
-                    logEntry.Labels[combinedProperty.Key] = combinedProperty.Value?.ToString() ?? "null";
+                    if (string.IsNullOrEmpty(combinedProperty.Key))
+                        continue;
+
+                    try
+                    {
+                        switch (Convert.GetTypeCode(combinedProperty.Value))
+                        {
+                            case TypeCode.Empty:
+                                propertiesStruct.Fields.Add(combinedProperty.Key, Value.ForNull());
+                                break;
+                            case TypeCode.Boolean:
+                                propertiesStruct.Fields.Add(combinedProperty.Key, Value.ForBool((bool)combinedProperty.Value));
+                                break;
+                            case TypeCode.Decimal:
+                            case TypeCode.Double:
+                            case TypeCode.Single:
+                            case TypeCode.Byte:
+                            case TypeCode.SByte:
+                            case TypeCode.Int16:
+                            case TypeCode.UInt16:
+                            case TypeCode.Int32:
+                            case TypeCode.UInt32:
+                            case TypeCode.Int64:
+                            case TypeCode.UInt64:
+                                propertiesStruct.Fields.Add(combinedProperty.Key, Value.ForNumber(Convert.ToDouble(combinedProperty.Value)));
+                                break;
+                            default:
+                                propertiesStruct.Fields.Add(combinedProperty.Key, Value.ForString(combinedProperty.Value?.ToString() ?? "null"));
+                                break;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        InternalLogger.Warn(ex, "GoogleStackdriver(Name={0}): Exception at BuildLogEntry with Key={1}", Name, combinedProperty.Key);
+                        propertiesStruct.Fields.Add(combinedProperty.Key, Value.ForNull());
+                    }
                 }
-                catch (Exception ex)
+
+                logEntry.JsonPayload = jsonStruct;

```

Comment:
By not adding the properties to `logEntry.Labels[]`, does this reduce the ability to filter/search using the StackDriver API or web UI?

Reply:
No idea. Not that experienced with Google StackDriver and how to perform optimal logging. I just tried to duplicate how the Serilog StackDriver was doing it.

Reply:
OK, they can always be added later.